### PR TITLE
Enhance blob upload support, to include metadata, and complete blob transfers (broadcast/private)

### LINKED ISSUE #75: BLOB transfer
Tracking the status of the asynchronous blob transfer feature, which still has some items remaining:

- [x] Data exchange plugin points (streaming optimized)
  - [x] Upload blob
  - [x] Download blob
  - [x] Transfer blob
  - [x] New blob received notification
- [x] HTTPS implementation
  - [x] Upload blob
  - [x] Download blob
  - [x] Transfer blob
  - [x] New blob received notification
- [x] Core API
  - [x] Upload blob
  - [x] Download blob
  - [x] Attach blob to broadcast message send
  - [x] Attach blob to private message send
- [x] Aggregator
  - [x] Do not mark message confirmed until blob data is available
  - [x] Rewind on arrival of blob data
- [x] e2e test
  - [x] Broadcast with blob data
  - [x] Private message with blob data

### PR DESCRIPTION
**Pull request by peterbroadhurst**
- Build now takes releases of UI, rather than building during the docker build
  - See https://github.com/hyperledger-labs/firefly-ui/pull/37
- Blobs now can have both a `blob` attachment and `value` JSON
  - `data.blob.hash` - the sha256 of the blob - must be there
  - `data.blob.public` - the public payload ref to download the blob - only if it's been published at some point
- On upload user can add some extra form parameters - must be before the upload file itself:
  - `metadata` - can be a string or a JSON object
  - `autometa` - if `true` then the `filename`,`size` and `mimetype` are added to JSON metadata
     - Can be combined with `metadata` if the metadata is JSON
  - `datatype.name` - if the metadata needs to be validated per a datatype
  - `datatype.version` - if the metadata needs to be validated per a datatype
  - `validator` - to customize the validator (only `json` is supported today)
- The `hash` of the `data` object created, will be one of:
  - The `value` hash - if just a JSON value
  - The `blob` hash reference - if just a BLOB value
  - A hash of the two HEX hashes concatenated together (no spaces or separators)
- `GET` `/api/v1/namespaces/:ns/data/:dataid/blob` to download blob
   - Some refactoring on the apiserver to support non-JSON `ReadCloser` results
- Broadcast blob transfers worked through
  - When sent in a broadcast message, the `data` will become pinned
  - The `data` has a `blob.public` reference added to it before being included in the batch
  - If it's sent multiple times, to multiple parties, the upload to public only happens once
- Public payload references move to a `string` (rather than `*Bytes32`)
  - **This updates the contract Solidity** - so this has a migration impact for CLI created envs
  - We were previously using a Bytes32 reference, derived from the IPFS ID. This was problematic
    - It was really confusing on the plugin interface, and for users
    - It would have got more confusing with the public reference on data
    - It requires the public storage plugin to be able to translate IDs into 32byte strings
       - This happens to be true for the majority of IPFS configurations today, but is not universal or extensible
- Private blob transfers worked through
  - Sends are initiated prior to the batch payload itself
  - Happen asynchronously, and have their own `operation` entry
- E2E tests included
  - Private blob transfer
  - Broadcast blob transfer
     
Fixes #75

### CODE REVIEW

**Comment by nguyer** [comment]
FYI @jebonfig this PR will require a new field added to the config to enable the UI:
```yaml
ui:
  enabled: true
```

**Review by nguyer — CHANGES_REQUESTED** [review]
One thing to check in the migrations, otherwise looks good.

File: db/migrations/ql/000022_create_blobs_table.up.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE blobs (
+  hash           string          NOT NULL,
+  payload_ref    string          NOT NULL,
+  created        int64           NOT NULL,
+  peer           string
```
[the remark this answers is not in the document]
Comment (peterbroadhurst) [reply db/migrations/ql/000022_create_blobs_table.up.sql:5]: Thanks @nguyer - this should be `NOT NULL` - I'll add a fix to that in #91

**Review by nguyer — APPROVED** [review]

**Comment by nguyer** [comment]
Marking approved, as the one issue I raised is fixed in the next PR

**Comment by peterbroadhurst** [comment]
Fixes https://github.com/hyperledger-labs/firefly/issues/77

**Pull request merged**